### PR TITLE
Fix pickling torch.float32

### PR DESCRIPTION
**Pull request by ailzhang**
Attempt fix for #14057 . This PR fixes the example script in the issue. 
The old behavior is a bit confusing here. What happened to pickling is python2 failed to recognize `torch.float32` is in module `torch`, thus it's looking for `torch.float32` in module `__main__`. Python3 is smart enough to handle it. 
According to the doc [here](https://docs.python.org/2/library/pickle.html#object.__reduce__), it seems `__reduce__` should return `float32` instead of the old name `torch.float32`. In this way python2 is able to find `float32` in `torch` module. 
> If a string is returned, it names a global variable whose contents are pickled as normal. The string returned by __reduce__() should be the object’s local name relative to its module

### CODE REVIEW

File: torch/csrc/Dtype.cpp [hunk 1]
```diff
@@ -52,7 +52,8 @@ static PyMethodDef THPDtype_methods[] = {
 
 PyObject *THPDtype_repr(THPDtype *self)
 {
-  return THPUtils_packString(self->name);
+  std::string name = self->name;
+  return THPUtils_packString("torch." + name);
```
Comment (driazati) [inline torch/csrc/Dtype.cpp]: If this is the full path of the dtype to search for it would break if someone did `import torch as tf`

**Review by ezyang — APPROVED** [review]
Can we get a test?

**Comment by ailzhang** [comment]
@BarclayII  @jermainewang @jackroos Could you help check if this PR solves your problems? I still want to get a confirmation from you guys before I merge it. Thanks!

**Pull request closed without merging**